anndata/experimental/merge.py

```diff
@@ -0,0 +1,602 @@
+def _(store: os.PathLike, *args, **kwargs) -> Union[ZarrGroup, H5Group]:
+    if store.suffix == ".h5ad":
+        import h5py
+
+        return h5py.File(store, *args, **kwargs)
+    import zarr
+
+    return zarr.open_group(store, *args, **kwargs)
+
+
+@as_group.register
+def _(store: str, *args, **kwargs) -> Union[ZarrGroup, H5Group]:
+    return as_group(Path(store), *args, **kwargs)
+
+
+@as_group.register(ZarrGroup)
+@as_group.register(H5Group)
+def _(store, *args, **kwargs):
+    return store
+
+
+###################
+# Reading
+###################
+
+
+def read_as_backed(group: Union[ZarrGroup, H5Group]):
+    """
+    Read the group until
+    SparseDataset, Array or EAGER_TYPES are encountered.
+    """
+
+    def callback(func, elem_name: str, elem, iospec):
+        if iospec.encoding_type in SPARSE_MATRIX:
+            return SparseDataset(elem)
+        elif iospec.encoding_type in EAGER_TYPES:
+            return read_elem(elem)
+        elif iospec.encoding_type == "array":
+            return elem
+        elif iospec.encoding_type == "dict":
+            return {k: read_as_backed(v) for k, v in elem.items()}
+        else:
+            return func(elem)
+
+    return read_dispatched(group, callback=callback)
+
+
+def _df_index(df: Union[ZarrGroup, H5Group]) -> pd.Index:
+    index_key = df.attrs["_index"]
+    return pd.Index(read_elem(df[index_key]))
+
+
+###################
+# Writing
+###################
+
+
+def write_concat_dense(
+    arrays: Sequence[Union[ZarrArray, H5Array]],
+    output_group: Union[ZarrGroup, H5Group],
+    output_path: Union[ZarrGroup, H5Group],
+    axis: Literal[0, 1] = 0,
+    reindexers: Reindexer = None,
+    fill_value=None,
+):
+    """
+    Writes the concatenation of given dense arrays to disk using dask.
+    """
+    import dask.array as da
+
+    def _gen_elem_to_append(elems):
+        for elem, ri in zip(elems, reindexers):
+            yield ri(elem, axis=1 - axis, fill_value=fill_value)
+
+    darrays = None
+    init_elem = arrays[0]
+
+    if isinstance(init_elem, H5Array):
+        darrays = (da.from_array(a, chunks="auto") for a in arrays)
+    elif isinstance(init_elem, ZarrArray):
+        darrays = (da.from_zarr(a) for a in arrays)
+
+    res = da.concatenate(
+        _gen_elem_to_append(
+            darrays,
+        ),
+        axis=axis,
+    )
+    write_elem(output_group, output_path, res)
+    output_group[output_path].attrs.update(
+        {"encoding-type": "array", "encoding-version": "0.2.0"}
+    )
+
+
+def write_concat_sparse(
+    datasets: Sequence[SparseDataset],
+    output_group: Union[ZarrGroup, H5Group],
+    output_path: Union[ZarrGroup, H5Group],
+    max_loaded_sparse_elems: int,
+    axis: Literal[0, 1] = 0,
+    reindexers: Reindexer = None,
+    fill_value=None,
+):
+    elems = _gen_slice_to_append(
+        datasets, reindexers, max_loaded_sparse_elems, axis, fill_value
+    )
+    init_elem = (csr_matrix, csc_matrix)[axis](next(elems))
```

Comment:
Does this load the first element into memory if it's not already?

Reply:
Yes, but they will be in memory; the problem is when I take a slice of the sparse dataset (i.e., read it partially), it sometimes returns it as other types of sparse format even if the SparseDataset itself has a determined format. I already checked if SparseDataset's are in the correct format and if they all have the same format. You can try to see this

```python
def write_concat_sparse(
    datasets: Sequence[SparseDataset],
    output_group: Union[ZarrGroup, H5Group],
    output_path: Union[ZarrGroup, H5Group],
    max_loaded_sparse_elems: int,
    axis: Literal[0, 1] = 0,
    reindexers: Reindexer = None,
    fill_value=None,
):
    elems = _gen_slice_to_append(
        datasets, reindexers, max_loaded_sparse_elems, axis, fill_value
    )
    init_elem = next(elems)
    write_elem(output_group, output_path, init_elem)
    del init_elem
    out_dataset: SparseDataset = read_as_backed(output_group[output_path])
    for temp_elem in elems:
        out_dataset.append(temp_elem)
        del temp_elem
```

I am still not sure why this happens I just used that as a workaround